common/lib/capa/capa/responsetypes.py

```diff
@@ -735,6 +801,92 @@ def get_score(self, student_answers):
 
+    def get_extended_hints(self, student_answers, new_cmap):
+        """
+        Extract compound and extended hint information from the xml based on the student_answers.
+        The hint information goes into the msg= in new_cmap for display.
+        Each choice in the checkboxgroup can have 2 extended hints, matching the
+        case that the student has or has not selected that choice:
+          <checkboxgroup label="Select the best snack" direction="vertical">
+             <choice correct="true">Donut
+               <choicehint selected="tRuE">A Hint!</choicehint>
+               <choicehint selected="false">Another hint!</choicehint>
+             </choice>
+        """
+        if self.answer_id in student_answers:
+            # Compound hints are a special thing just for checkboxgroup, trying
+            # them first before the regular extended hints.
+            if self.get_compound_hints(new_cmap, student_answers):
+                return
+
+            # Look at all the choices - each can generate some hint text
+            choices = self.xml.xpath('//checkboxgroup[@id=$id]/choice', id=self.answer_id)
+            hint_divs = ''
+            label = None
+            label_count = 0
+            # We build up several hints in hint_divs, then wrap it once at the end.
+            for choice in choices:
+                name = choice.get('name')  # generated name, e.g. choice_2
+                if name in student_answers[self.answer_id]:
+                    selector = 'true'  # true/false attribute used in the hint xml
+                else:
+                    selector = 'false'
+                # We find the matching <choicehint> in python vs xpath so we can be case-insensitive
+                hint_nodes = choice.findall('./choicehint')
+                for hint_node in hint_nodes:
+                    if hint_node.get('selected', '').lower() == selector:
+                        text = hint_node.text.strip()
+                        if hint_node.get('label') is not None:  # tricky: label '' vs None is significant
+                            label = hint_node.get('label')
+                            label_count += 1
+                        if text:
+                            # Unusual: there can be multiple hint divs across the choices, all cat'd together
+                            hint_divs += '<div class="{0}">{1}</div>'.format(QUESTION_HINT_TEXT_STYLE, text)
+                        break
+            if hint_divs:
+                # Complication: if there is only a single label specified, we use it.
+                # However if there are multiple, we use none.
+                if label_count > 1:
+                    label = None
+                new_cmap[self.answer_id]['msg'] += (
+                    self.make_hint_div(None,
+                                       new_cmap[self.answer_id]['correctness'] == 'correct',
+                                       label, hint_divs))
```

Comment:
Hmm, I seem to have a problem writing the line-breaking per style correctly. Reading the style page again, I now see how this can actually should be done without the extra ( .. )